### PR TITLE
Get diet requirement from database

### PR DESCRIPTION
**Pull request by 090997**
resolve: #17 

### Short Description 
we have a new pages for Volunnteer that they can choose more option about their diet_requirement.So we create a  table to store these data. And we need to get the data about diet_requirement from database.

### Implementation

- Create a request parser (get_parser) for parsing user_id from the incoming GET request.
- Define the diet_requirement_fields dictionary to map the properties of the dietary requirement to appropriate data types.
- Create a Flask Blueprint named diet_requirement_retrieval_bp and an Api instance associated with the blueprint.
- Add the RetrieveDietaryRequirement resource to the API with the endpoint /getDietaryRequirement.

### Testing cases:
write two cases to test whether it get the correct answer from database

- [ ] **test case 1:**

            diet_requirement1 = DietRequirement(
            user_id=2,
            diet_id=2,
            halal=False,
            vegetarian=True,
            vegan=False,
            nut_allergy=False,
            shellfish_allergy=False,
            gluten_intolerance=True,
            kosher=False,
            lactose_intolerance=True,
            diabetic=True,
            egg_allergy=False,
            other="without oil"
        )}

- [ ] **test 2**

        # Test Case 2
        diet_requirement2 = DietRequirement(
            user_id=3,
            diet_id=3,
            halal=False,
            vegetarian=False,
            vegan=True,
            nut_allergy=False,
            shellfish_allergy=True,
            gluten_intolerance=False,
            kosher=False,
            lactose_intolerance=False,
            diabetic=False,
            egg_allergy=False,
            other=""
        )

- [ ] test 3
        input user_id where is not in database
        user_id=1234

**Test result**

 
test 1 return
Return correct diet_requirement and return true
image

test 2 return
Return correct diet_requirement and return true
image

test 3 return
Return false
image

### CODE REVIEW

File: application.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 from flask import Flask
 from flask_cors import CORS
 from controllers import *
+from controllers.change_diet_requirement import diet_requirement_change_bp
```
Comment (purple-waves) [inline application.py:4]: - This import should not be necessary, it should be covered by the previous line. 
import * means import all

File: controllers/__init__.py [hunk 2]
```diff
@@ -16,10 +16,13 @@
 from .user_type import user_type_bp
 from .tenancy_config import tenancy_config_bp
 from .tutorial_quiz import tutorial_quiz_bp
-from .email import email_bp
+from .email_controllers import email_bp
```
Comment (purple-waves) [inline controllers/__init__.py:19]: Why is this being renamed?

File: controllers/__init__.py [hunk 3]
```diff
@@ -16,10 +16,13 @@
...
 from .volunteer_unavailability import volunteer_unavailability_bp
 from .user import user_bp
 from .chatbot import chatbot_bp
 from .profile import profile_bp
+from .diet_requirement_retrieval import diet_requirement_retrieval_bp
+
+
```
Comment (purple-waves) [inline controllers/__init__.py:26]: Do these blank lines need to be here?

File: domain/__init__.py [hunk 4]
```diff
@@ -3,3 +3,4 @@
 from .type import *
 from .entity import *
 
+
```
Comment (purple-waves) [inline domain/__init__.py:6]: Remove the addition of this line. 

Remember to double check the changes you make, we want to keep the changes relevant to the issue they're fixing.

File: domain/base.py [hunk 5]
```diff
@@ -12,7 +12,7 @@
 
 # Configure Session
 Session = sessionmaker()
-Engine = create_engine('mysql+mysqldb://{0}:{1}@{2}:{3}/{4}'.format(secret.get()['username'],
+Engine = create_engine('mysql+mysqlconnector://{0}:{1}@{2}:{3}/{4}'.format(secret.get()['username'],
```
Comment (purple-waves) [inline domain/base.py:15]: Why did this need to change?

File: domain/entity/__init__.py [hunk 6]
```diff
@@ -11,4 +11,4 @@
 from .question import Question
 from .password_retrieval import PasswordRetrieval
 from .unavailability_time import UnavailabilityTime
-from .chatbot_input import ChatbotInput
\ No newline at end of file
+from .chatbot_input import ChatbotInput
```
Comment (purple-waves) [inline domain/entity/__init__.py:14]: There shouldn't be a change here. Maybe you've put in a character from a keyboard in another language?

See if you can revert changes to this file.

File: domain/entity/diet_requirement.py [hunk 7]
```diff
@@ -4,6 +4,8 @@
 from domain import Base
 
 
+
+
```
Comment (purple-waves) [inline domain/entity/diet_requirement.py:8]: Are these extra lines required?

File: domain/entity/diet_requirement_change.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
+
```
Comment (purple-waves) [inline domain/entity/diet_requirement_change.py:1]: There shouldn't be a blank line here

File: repository/change_diet_repository.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
+
+
```
Comment (purple-waves) [inline repository/change_diet_repository.py:2]: Remove blank lines

File: repository/change_diet_repository.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+Update the dietary requirements of a user.
+
+Parameters:
+
+session: database session
+user_id: ID of the user
+args: parameters containing dietary requirements
+Returns:
+
+True if the update is successful; False otherwise.
```
Comment (purple-waves) [inline repository/change_diet_repository.py:16]: The indentation should be the same as the code it's for (and the """). Please indent this section

File: repository/change_diet_repository.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    except Exception as e:
+        print(f"Error updating dietary requirements: {e}")
+        return False
+
+def get_user_by_id(session, user_id):
+    """
+Find user by ID
+:param session: database session
+:param user_id: ID of the user
+:return: if user is found, return the user object, otherwise return None
```
Comment (purple-waves) [inline repository/change_diet_repository.py:33]: Indent please

**Review by purple-waves — CHANGES_REQUESTED** [review]
I've left a heap of comments.
Some of the comments request changes, some of them are questions that I don't know the answer to. Please reply to the comments if you have questions about them, and also reply to the comments with questions in them.

**Comment by purple-waves** [comment]
The pull request description doesn't need to have implementation details, that's all included in the code.

File: application.py [hunk 12]
```diff
@@ -52,3 +59,5 @@ def main():
 
     logging.basicConfig(filename='error.log', level=logging.DEBUG)
     app.run(host='0.0.0.0')
+
+
```
Comment (purple-waves) [inline application.py:63]: You only need one blank line at the end of the file

File: controllers/change_diet_requirement.py
Comment (purple-waves) [inline controllers/change_diet_requirement.py:1]: The change diet requirement file is not necessary. We only need to implement api endpoints for storing requirements, which Rita has implemented in #16, and getting requirements, which is what you have done in diet_requirement_retrieval. This means that this file should be deleted.

File: controllers/diet_requirement_retrieval.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+from domain import session_scope
+
+from repository.diet_requirement_repository import get_dietary_requirements
+
+# Request parser
+get_parser = reqparse.RequestParser()
+get_parser.add_argument('user_id', type=int, required=True)
+
+# DietRequirement fields
```
Comment (purple-waves) [inline controllers/diet_requirement_retrieval.py:12]: Comments are generally good, but in this case the comment doesn't add anything that the code doesn't tell us already. The same applies for the "Request Parser" comment above and "Result fields" comment below. Please remove them.

File: controllers/test_get_diet_requirement_retriveal.py
Comment (purple-waves) [inline controllers/test_get_diet_requirement_retriveal.py:1]: Some useful test cases, but they should be in postman rather than the codebase. If this file only exists to run test cases, then it should be removed.

File: domain/entity/diet_requirement_change.py
Comment (purple-waves) [inline domain/entity/diet_requirement_change.py:1]: Is this any different from domain/entity/diet_requirement.py? If not, then that file should be used instead, and this one removed.

**Review by purple-waves — CHANGES_REQUESTED** [review]
Some more comments.

File: controllers/__init__.py (hunk 3 above)
Comment (090997) [reply controllers/__init__.py:26]: no, it can be deleted

File: controllers/__init__.py (hunk 2 above)
Comment (090997) [reply controllers/__init__.py:19]: i think i  didin't change anything about these two line.

Comment (purple-waves) [reply controllers/__init__.py:19]: Ok, they got changed somehow, so you'll need to change them back. I think you should be able to just change that line to `from .email import email_bp` and that'll fix it

File: controllers/diet_requirement_retrieval.py [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+                diet_requirement = get_dietary_requirements(session, user_id)
+
+            return {"diet_requirement": diet_requirement, "success": True}
+        except Exception as e:
+            print(f"Error: {e}")
+            return {"success": False}, 400
+
+
+
+
```
Comment (purple-waves) [inline controllers/diet_requirement_retrieval.py:60]: There should only be 2 blank lines here

File: controllers/test_get_diet_requirement_retriveal.py
Comment (purple-waves) [inline controllers/test_get_diet_requirement_retriveal.py:1]: This file shouldn't be included in the pull request, we want to keep testing entirely within postman for now.

File: domain/entity/diet_requirement.py
Comment (purple-waves) [inline domain/entity/diet_requirement.py:1]: There aren't any functional changes in this file, so these changes shouldn't be in the pull request either.

**Review by purple-waves — CHANGES_REQUESTED** [review]
This is looking a lot better, thanks for all your changes so far. Just a few more to go

File: controllers/email_controllers.py
Comment (purple-waves) [inline controllers/email_controllers.py:1]: This rename needs to be reversed. At the moment, the api crashes as soon as you access it

**Review by purple-waves — CHANGES_REQUESTED** [review]

File: domain/entity/diet_requirement.py
Comment (purple-waves) [reply domain/entity/diet_requirement.py:1]: I don't think its going to be worth trying to change these back, but remember to double check the changes you're committing in future pull requests.

File: controllers/diet_requirement_retrieval.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    """
+    This is a class to retrieve the dietary requirement data from the database
+    """
+
+    @marshal_with(result_fields)
+    def get(self):
+        try:
+            print("RetrieveDietaryRequirement.get() called")
+            user_id = request.args.get('user_id', None, type=int)
+            if user_id is None:
```
Comment (purple-waves) [inline controllers/diet_requirement_retrieval.py:45]: Thanks for adding this.

File: controllers/diet_requirement_retrieval.py [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+class RetrieveDietaryRequirement(Resource):
+    """
+    This is a class to retrieve the dietary requirement data from the database
+    """
+
+    @marshal_with(result_fields)
+    def get(self):
+        try:
+            print("RetrieveDietaryRequirement.get() called")
```
Comment (purple-waves) [inline controllers/diet_requirement_retrieval.py:43]: We should try to remove print statements before pushing to main, unless they are still needed for something. In this case, I don't think the print statement is still necessary so it should be removed.

File: repository/diet_requirement_repository.py [hunk 17]
```diff
@@ -0,0 +1,8 @@
+from domain.entity.diet_requirement import DietRequirement
+
+
+def get_dietary_requirements(session, user_id):
+    diet_requirement = session.query(DietRequirement).filter(DietRequirement.user_id == user_id).first()
+    if diet_requirement:
+        session.expunge(diet_requirement)
```
Comment (purple-waves) [inline repository/diet_requirement_repository.py:7]: I'm not familiar with session.expunge, why is it required here?

**Review by purple-waves — CHANGES_REQUESTED** [review]
So close! One more style thing and a clarification

File: controllers/diet_requirement_retrieval.py [hunk 18]
```diff
@@ -50,7 +49,6 @@ def get(self):
 
             return {"diet_requirement": diet_requirement, "success": True}
         except Exception as e:
-            print(f"Error: {e}")
```
Comment (purple-waves) [inline controllers/diet_requirement_retrieval.py]: Thanks for taking this one out, I hadn't noticed it.

File: repository/diet_requirement_repository.py
Comment (purple-waves) [inline repository/diet_requirement_repository.py:1]: Sorry I didn't notice this before, but this function should really be in `repository/diet_repository`. We want to keep all functions for manipulating the dietary requirements data in one place.

**Review by purple-waves — CHANGES_REQUESTED** [review]

File: controllers/diet_requirement_retrieval.py [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+                diet_requirement = get_dietary_requirements(session, user_id)
+
+            return {"diet_requirement": diet_requirement, "success": True}
+        except Exception as e:
+            return {"success": False}, 400
+
+
+diet_requirement_retrieval_bp = Blueprint('diet_requirement_retrieval', __name__)
+api = Api(diet_requirement_retrieval_bp, '/dietary')
+api.add_resource(RetrieveDietaryRequirement, '/getDietaryRequirement')
```
Comment (BenMMcLean) [inline controllers/diet_requirement_retrieval.py:57]: Could we change this to follow proper REST convention (change it to just "/requirements")

**Review by BenMMcLean — CHANGES_REQUESTED** [review]
Just my two cents in addition to Rachel's feedback

**Review by BenMMcLean — APPROVED** [review]
